bundle/config/mutator/python/python_locations.go

```diff
@@ -0,0 +1,163 @@
+package python
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"path/filepath"
+
+	"github.com/databricks/cli/libs/dyn"
+)
+
+const generatedFileName = "__generated_by_pydabs__.yml"
+
+// pythonLocations is data structure for efficient location lookup for a given path
+type pythonLocations struct {
+	// descendants referenced by index, e.g. '.foo'
+	keys map[string]*pythonLocations
+
+	// descendants referenced by key, e.g. '[0]'
+	indexes map[int]*pythonLocations
+
+	// location for the current node if it exists
+	location dyn.Location
+
+	// if true, location is present
+	exists bool
+}
+
+type pythonLocationEntry struct {
+	Path   string `json:"path"`
+	File   string `json:"file"`
+	Line   int    `json:"line"`
+	Column int    `json:"column"`
+}
+
+// mergePythonLocations applies locations from Python mutator into given dyn.Value
+//
+// The primary use-case is to merge locations.json with output.json, so that any
+// validation errors will point to Python source code instead of generated YAML.
+func mergePythonLocations(value dyn.Value, locations *pythonLocations) (dyn.Value, error) {
+	return dyn.Walk(value, func(path dyn.Path, value dyn.Value) (dyn.Value, error) {
+		if newLocation, ok := findPythonLocation(locations, path); ok {
+			var newLocations []dyn.Location
+
+			// the first item in the list is the "last" location used for error reporting
+			newLocations = append(newLocations, newLocation)
+
+			for _, location := range value.Locations() {
+				if filepath.Base(location.File) == generatedFileName {
+					continue
+				}
+
+				// don't add duplicates if location already exists
+				if location == newLocation {
```

Comment:
When can this happen?

Reply:
It can't really happen with the current code path. Location array will always contain a single element with a virtual path, but I think it makes sense from the behavior not to add duplicates. However, since we don't really need it, I will just remove it because it's less code to have.